Fixed #27295 -- Added a system check to prohibit model names that start/end with an underscore or contain double underscores.

inondle opened the pull request:
https://code.djangoproject.com/ticket/27295

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -155,3 +155,30 @@ def build_error(model_key, func, args, keywords):
+    if app_configs is None:
+        models = apps.get_models()
+    else:
+        models = chain.from_iterable(app_config.get_models() for app_config in a
```
This line isn't complete.

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -155,3 +155,30 @@ def build_error(model_key, func, args, keywords):
+        if model_name.startswith('_') or model_name.endswith('_'):
+            errors.append(
+                Error(
+                    "The model name '%s' starts or ends with underscores." % mod
```
This line is truncated.

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -155,3 +155,30 @@ def build_error(model_key, func, args, keywords):
+        elif '__' in model_name:
+            errors.append(
+                Error(
+                    "The model name '%s' contains double underscores." % model_n
```
This line is truncated.

timgraham commented on docs/ref/checks.txt:
```diff
@@ -141,6 +141,9 @@ Models
   doesn't provide model ``<model>``.
+* **models.E023**: The model name ``<model>`` starts or ends with underscores.
+* **models.E024**: The model name ``<model>`` contians double underscores.
+
```
remove blank line

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -155,3 +155,30 @@ def build_error(model_key, func, args, keywords):
 @register(Tags.models)
 def check_lazy_references(app_configs=None, **kwargs):
     return _check_lazy_references(apps)
+
```
Add a blank line above (check flake8)

pope1ni commented on docs/ref/checks.txt:
```diff
@@ -141,6 +141,9 @@ Models
   ``<app label>>.<model>``, but app ``<app label>`` isn't installed or
   doesn't provide model ``<model>``.
+* **models.E023**: The model name ``<model>`` starts or ends with underscores.
+* **models.E024**: The model name ``<model>`` contians double underscores.
```
`contians` → `contains`

pope1ni replied:
Also, perhaps it is worth saying why this is not allowed, i.e. double underscores are the lookup / join separator.

pope1ni commented on django/core/checks/model_checks.py:
```diff
@@ -155,3 +155,30 @@ def build_error(model_key, func, args, keywords):
+                    id='models.E023'
+                )
+            )
+        elif '__' in model_name:
```
Please use `django.db.models.constants.LOOKUP_SEP`

inondle commented:
I cleaned up the typos and added a blurb to the docs explaining why you can't name models this way. Please let me know of any changes I need to make. 

 I'm still perplexed by the test issue. I'm still working on it but does anyone see anything glaring that I overlooked with the tests?

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -7,6 +7,7 @@
 
 from django.apps import apps
 from django.core.checks import Error, Tags, register
+from django.db.models import constants
```
I think using `from django.db.models.constants import LOOKUP_SEP` will fix the circular import issue causing some test failures.

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -155,3 +156,31 @@ def build_error(model_key, func, args, keywords):
     return _check_lazy_references(apps)
+
+
+@register(Tags.models)
```
The check should probably go in `django/db/models/base.py`. As registered, this isn't called by `Model.check()`.

timgraham commented on docs/ref/checks.txt:
```diff
@@ -141,6 +141,13 @@ Models
+* **models.E024**: The model name ``<model>`` contians double underscores.
+  This can be mistaken for the database layer's ``join`` and ``lookup``
+  separator.
+
```
chop blank line (just 1 needed between sections)

timgraham commented on docs/ref/checks.txt:
```diff
@@ -141,6 +141,13 @@ Models
 * **models.E022**: ``<function>`` contains a lazy reference to
   ``<app label>>.<model>``, but app ``<app label>`` isn't installed or
   doesn't provide model ``<model>``.
+* **models.E023**: The model name ``<model>`` starts or ends with underscores.
```
"... cannot start or end with an underscore as it collides with the relational lookup syntax." (or something like that)
The message here should be the same as in the check.

timgraham commented on tests/invalid_models_tests/test_models.py:
```diff
@@ -689,6 +689,49 @@ class Meta:
+            )
+        ]
+        self.assertEqual(errors, expected)
+
```
chop extra blank line (please check flake8)

inondle commented:
Hi guys, sorry for the long wait. I've been dealing with midterms this past couple weeks. I have the checks implemented in the `django/db/models/base.py` file and have my checks running correctly.

The issue now is that my tests cause an different test to fail. I think it might be trying to operate on a model that contains the error I'm trying to fix. The test is in `test/invalid_models_tests/test_relative_fields.test_related_field_has_valid_related_name`. This test seems to run through a series of possible model names and make sure that none of them give any errors. The issue is that the related name `_starts_with_underscore'  results in the model name `Child__starts_with_underscore`. This model name triggers the new system check and returns an error causing the test to fail. I'm not sure what to do in this case, you guys have any ideas?

timgraham commented:
It looks like you need the test just needed updating. What was tested as valid is actually invalid. I've updated the branch and will give it another review later.
